p2p/host/basic/basic_host.go

```diff
@@ -377,9 +390,43 @@ func (h *BasicHost) Connect(ctx context.Context, pi pstore.PeerInfo) error {
 		return nil
 	}
 
+	resolved, err := h.resolveAddrs(ctx, h.Peerstore().PeerInfo(pi.ID))
+	if err != nil {
+		return err
+	}
+	h.Peerstore().AddAddrs(pi.ID, resolved, pstore.TempAddrTTL)
+
 	return h.dialPeer(ctx, pi.ID)
 }
 
+func (h *BasicHost) resolveAddrs(ctx context.Context, pi pstore.PeerInfo) ([]ma.Multiaddr, error) {
+	proto := ma.ProtocolWithCode(ma.P_IPFS).Name
+	p2paddr, err := ma.NewMultiaddr("/" + proto + "/" + pi.ID.Pretty())
+	if err != nil {
+		return nil, err
+	}
+
+	var addrs []ma.Multiaddr
+	for _, addr := range pi.Addrs {
+		addrs = append(addrs, addr)
+
+		reqaddr := addr.Encapsulate(p2paddr)
+		resaddrs, err := h.maResolver.Resolve(ctx, reqaddr)
```

Comment:
> Curious, what would try to match-check here?

Oh I see, whether the returned address's peerid matches the input peerid -- the resolver takes care of that. It's pretty neat, see https://github.com/multiformats/go-multiaddr-dns readme for examples, in particular the /dnsaddr examples. For /dns4 and /dns6 there's nothing to match-check since these only do A/AAAA lookups.